File: Class_11_Linked_List_Problems_Medium/inClassWork/linkedNodeInClassExample.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert(self, data):
        newNode = Node(data)
        if(self.head):
            current = self.head
            while(current.next):
                current = current.next
            current.next = newNode
        else:
            self.head = newNode

    def remove(self, data):
        if data == self.head.data:
            self.head = self.head.next
        else:
            current = self.head.next
            prev = self.head
            while(current):
                if current.data == data:
                    prev.next = current.next
                    break
                prev = current
                current = current.next

    def removeDups(self):
        current = self.head.next
        prev = self.head
        while(current):
            if current.data == prev.data:
                prev.next = current.next
                current = current.next
            else:
                prev = current
                current = current.next




def printHead(head):
    current = head
    while(current):
        print(current.data)
        current = current.next

def reverse(node):
    if (node == None):
        return node
    if (node.next == None):
        return node

    tempnode = reverse(node.next)
    node.next.next = node
    node.next = None
    return tempnode


LL = LinkedList()
LL.insert(3)
LL.insert(3)
LL.insert(3)
LL.insert(4)
LL.insert(5)
LL.insert(5)
LL.insert(6)
printHead(LL.head)
print()
LL.removeDups()
printHead(LL.head)
print()
head = reverse(LL.head)
printHead(head)

File: Class_11_Linked_List_Problems_Medium/inClassWork/linkedNodeInClassExample.js
class Node {
  constructor(data) {
    this.data = data
    this.next = null
  }
}

function initializeLinkedList(array) {
  let node = new Node(array.shift())
  if (array.length) {
    node.next = initializeLinkedList(array)
  }
  return node
}
let head = initializeLinkedList([1, 1, 3, 5, 5]);


function reverse(head) {
  let prev = null
  let current = head

  while (current !== null) {
    let next = current.next
    current.next = prev
    prev = current
    current = next
  }
  return prev
}

function deleteNode(val) {
  if (this.head.data === val) {
    this.head = this.head.next
    return
  }
  let prev = null
  let current = this.head
  while (current !== null) {
    if (current.data === val) {
      prev.next = current.next
      break
    }
  }
}

function removeDuplicates(head) { // not finished in class loop though array and keep all numbers, then filters array to just have the duplicates
  let prev = null
  let current = head
  let duplicates = []

  while (current !== null) {
    duplicates.push(current.data)
    let next = current.next
    prev = current
    current = next
  }
}

function deleteNnodesAfterMnodes(head, n, m) { // not finished in class loop though array and keep all numbers, then filters array to just have the duplicates
  let prev = null
  let current = head

  while (current !== null) {
    let next = current.next
    prev = current
    current = next
  }
}

deleteNnodesAfterMnodes(head, 1, 2)
//
// return arr.filter((item,
//                    index) => arr.indexOf(item) === index);
// }